syntax = "proto3";

package biatob.proto.mvp;

// For MVP purposes, we're just gonna store the entire worldstate in a protobuf blob. "Data base"? Huh?
// TODO(P3): ^ stop doing that
message WorldState {

  map<uint64, UserInfoTodoUnclash> users = 1;
  map<uint64, Market> markets = 2;

  map<string, uint64> auth_token_owner_ids = 3;

  message UserInfoTodoUnclash {
    string username = 1;
    bytes password_bcrypt = 2;
    repeated uint64 trusted_users = 3;
    // TODO: string email = 3; // optional
  }
  message Market {
    // Set upon creation, never changed
    string question = 1;
    CertaintyRange certainty = 2;
    uint64 maximum_stake_cents = 3;
    uint64 created_unixtime = 4;
    uint64 closes_unixtime = 5;
    string special_rules = 6;
    uint64 creator_id = 7;

    // Changes regularly
    repeated Trade trades = 8;
    Resolution resolution = 9;
  }

  message Trade {
    uint64 bettor_id = 1;
    bool expected_resolution = 2;
    uint64 bettor_stake = 3;
    uint64 transacted_unixtime = 4;
  }
}


enum Void {
  VOID = 0;
}

enum Resolution {
  RESOLUTION_NONE_YET = 0;
  RESOLUTION_YES = 1;
  RESOLUTION_NO = 2;
}

message Position {
  int64 win_cents_if_yes = 1;
  int64 win_cents_if_no = 2;
}

message Auth {
  oneof auth_kind {
    string magic_token = 1;
  }
}
message AuthError {
  oneof auth_error_kind {
    Void invalid_token = 1;
  }
}

message SignUpRequest {
  string email = 1;
  string password = 2;
  string display_name = 3;
}
message SignUpResponse {
  oneof signup_result {
    Void ok = 1;
    Error error = 2;
  }
  message Error {
    string catchall = 1;
    Void email_already_registered = 2;
  }
}

message CertaintyRange {
  float low = 1;
  float high = 2;
}
message MarketPrivacy {
  oneof privacy_kind {
    Void all_trusted_by_author = 1;
    Emails specific_users = 2;
  }
  message Emails {
    repeated string emails = 1;
  }
}
message CreateMarketRequest {
  Auth auth = 1;
  string question = 2;
  MarketPrivacy privacy = 3;
  CertaintyRange certainty = 4;
  uint64 maximum_stake_cents = 5;
  uint64 open_seconds = 6;
  string special_rules = 7;
}
message CreateMarketResponse {
  oneof create_market_result {
    uint64 new_market_id = 1;
    Error error = 2;
  }
  message Error {
    string catchall = 1;
    AuthError auth_error = 2;
  }
}

message GetMarketRequest {
  Auth auth = 1;
  uint64 market_id = 2;
}
message GetMarketResponse {
  oneof get_market_result {
    Market market = 1;
    Error error = 2;
  }
  message Market {
    string question = 1;
    CertaintyRange certainty = 2;
    uint64 maximum_stake_cents = 3;
    uint64 remaining_yes_stake_cents = 4;
    uint64 remaining_no_stake_cents = 5;
    uint64 created_unixtime = 6;
    uint64 closes_unixtime = 7;
    string special_rules = 8;
    UserInfo creator = 9;
    Resolution resolution = 10;
  }
  message Error {
    string catchall = 1;
    AuthError auth_error = 2;
  }
}
message UserInfo {
  string display_name = 1;
}

message StakeRequest {
  Auth auth = 1;
  uint64 market_id = 2;
  bool expected_resolution = 3;
  uint64 stake = 4;
}
message StakeResponse {
  oneof stake_result {
    Void ok = 1;
    Error error = 2;
  }
  message Error {
    string catchall = 1;
    AuthError auth_error = 2;
  }
}

message GetUserRequest {
  Auth auth = 1;
  string email = 2;
}
message GetUserResponse {
  oneof get_user_result {
    User user = 1;
    Error error = 2;
  }
  message User {
    bool trusted_by_requester = 1;
    bool trusts_requester = 2;
  }
  message Error {
    string catchall = 1;
    AuthError auth_error = 2;
  }
}

message MarkTrustedRequest {
  Auth auth = 1;
  string email_to_trust = 2;
}
message MarkTrustedResponse {
  oneof result {
    Void ok = 1;
    Error error = 2;
  }
  message Error {
    string catchall = 1;
    AuthError auth_error = 2;
  }
}
